### PR TITLE
Saving checkpoints periodically for preference comparisons

### PR DESCRIPTION
**Pull request by kmdanielduan**
This PR aims to
- Allow additional self-defined `callback` to happen in `main_trainer.train()`.
- Allow periodical saving of reward model and policy weights at the end of `main_trainer.train()`.

### CODE REVIEW

File: src/imitation/algorithms/preference_comparisons.py [hunk 1]
```diff
@@ -812,7 +816,14 @@ def __init__(
 
         self.dataset = PreferenceDataset()
 
-    def train(self, total_timesteps: int, total_comparisons: int) -> Mapping[str, Any]:
+        self.vec_normalize = vec_normalize
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:819]: This attribute doesn't seem to actually be used anywhere?

File: src/imitation/scripts/train_preference_comparisons.py [hunk 2]
```diff
@@ -24,6 +24,20 @@
...
 
+def save(trainer, save_path):
+    """Save reward model and policy."""
+    os.makedirs(save_path, exist_ok=True)
+    th.save(trainer.reward_trainer.model, os.path.join(save_path, "reward_net.pt"))
+    if hasattr(trainer.trajectory_generator, "algorithm"):
+        serialize.save_stable_model(
+            os.path.join(save_path, "policy"),
+            trainer.trajectory_generator.algorithm,
+            trainer.vec_normalize,
```
Comment (AdamGleave) [inline src/imitation/scripts/train_preference_comparisons.py:35]: Why not have `save` take `vec_normalize` directly? It's created inside the script, and not used by the trainer class at all.

File: src/imitation/algorithms/preference_comparisons.py (hunk 1 above)
Comment (kmdanielduan) [reply src/imitation/algorithms/preference_comparisons.py]: You are right about it!

File: src/imitation/scripts/train_preference_comparisons.py (hunk 2 above)
Comment (kmdanielduan) [reply src/imitation/scripts/train_preference_comparisons.py]: Good point. I missed out that previously.

File: src/imitation/scripts/config/train_preference_comparisons.py [hunk 3]
```diff
@@ -38,6 +38,8 @@ def train_defaults():
     normalize = True  # Use VecNormalize
     normalize_kwargs = {"norm_reward": False}  # kwargs for `VecNormalize`
 
+    checkpoint_interval = 0  # Num epochs between checkpoints (<0 disables)
```
Comment (AdamGleave) [inline src/imitation/scripts/config/train_preference_comparisons.py:41]: `=0` disables all but final one right (which I'm not sure I'd call a "checkpoint" anyway)?

```suggestion
    checkpoint_interval = 0  # Num epochs between checkpoints (<0 disables, =0 final only)
```

File: src/imitation/scripts/train_preference_comparisons.py [hunk 4]
```diff
@@ -24,6 +24,38 @@
...
+    os.makedirs(save_path, exist_ok=True)
+    th.save(trainer.reward_trainer.model, os.path.join(save_path, "reward_net.pt"))
+    if allow_save_policy:
+        # Note: We should only save the model as model.pkl if `trajectory_generator`
+        # contains one. Specifically we check if the `trajectory_generator` contains an
+        # `algorithm` attribute.
+        assert hasattr(trainer.trajectory_generator, "algorithm")
+        save_model(trainer.trajectory_generator, vec_normalize, save_path)
+    else:
+        print("trainer.trajectory_generator doesn't contain a policy to save.")
```
Comment (AdamGleave) [inline src/imitation/scripts/train_preference_comparisons.py:56]: Avoid print statements; use `custom_logger.warn` instead (it'll then get logged to files not just stdout)

**Review by AdamGleave — APPROVED** [review]
LGTM apart from one minor comment.

File: src/imitation/scripts/train_preference_comparisons.py (hunk 4 above)
Comment (kmdanielduan) [reply src/imitation/scripts/train_preference_comparisons.py:56]: Good point! I would log the warning messages to trainer.logger.

**Pull request merged**